### ISSUE TITLE
How to enumerate all classes defined by the JDK / How to parse the "modules" file

### ISSUE THREAD

**Issue by stefan-reich** [eclipse-openj9/openj9#15537]
Hi, for a scripting language implemented in Java I need to make a list of all classes defined in the current JDK at runtime.

The way I found to do this for a regular JDK is by inspecting the .jmod files in the JDK directory (which are really just jar files, so it's not hard to parse them).

Issue is, this doesn't work with OpenJ9 as it doesn't have .jmod files.

Instead it has the "modules" file which I'm totally not sure how to parse. (Magic bytes DADAFECA. Doesn't ring any bell...)

So how can I do this with OpenJ9? Thanks

**Comment by pshipton** [comment]
OpenJ9 also has .jmod files, assuming you are using jdk11 or later. Are you looking at a JRE build rather than a JDK build?

The modules file can be opened using the `jimage` utility, although that utility is only part of the JDK build.
``bin/jimage` list lib/modules` provides a list of the classes.
```
Usage: jimage <extract | info | list | verify> <options> jimage...
use -h or --help for a list of possible options.
```

**Comment by pshipton** [comment]
btw, by looking at the `jimage` implementation code, you can see how to read the modules file using internal APIs (`jdk.internal.jimage`). These classes/APIs are internal and need to be exposed via command line options. Since they are internal they are subject to change without notice, as is the format of the modules file itself.
https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/openj9/src/jdk.jlink/share/classes/jdk/tools/jimage/JImageTask.java

**Comment by stefan-reich** [comment]
Hi @pshipton Thank you very much, this answers my questions. So I basically have to tell users not to use OpenJ9 JRE (but OpenJ9 JDK will be fine). Shouldn't be a problem for anyone.
